---
title: "eDNA"
author: "Chris Kracha"
date: "06/30/2022"
output: html
---

This document contains cleaning scripts for eDNA collected in 2021-2022 through JonahVentures. Tables from JV are reformatted to fit our needs, factors are reordered and relabeled, additional metadata is created, and all rounds of sampling are combined. Products of this document include long and wide tables, as well as summary tables that make the data more accessible.

Note that raw_data/eDNA_sample_ids.csv needs to be manually updated with sample information for future additions of data.

```{r setup}
knitr::opts_chunk$set(message = FALSE)
install.packages("ggforce")
library(taxa) # special taxa data class
library(taxize)
library(rentrez) # data base connections for common names
library(metacoder) # taxonomy data parsing
library(janitor)
library(lubridate)
library(tidyverse)
library(dplyr)
library(ggforce)
```

# Write cleaning function
```{r clean batch eDNA}
# The first three samples were stored in alcohol and did the DNA was degraded. 
# Metadata for these samples was added manually and dates are approximate.

# Function cleans and combines all eDNA files from a folder
eDNA_clean <- function(folder) {
  
  cleaned <- folder %>%
    
    # Create list of all files in target folder
    list.files(., full.names = TRUE)  %>%
    
    # Read CSV files and bind rows together
    ldply(read_csv) %>% 
    
    # Clean names
    dplyr::rename(percent_match = "% match",
           number_species = "# species") %>%
    
    # Combine taxa into one column
    unite("kpcofgs", Kingdom:Species, 
          sep = ", ", na.rm = TRUE, 
          remove = FALSE) %>%
    
    select(1:number_species, kpcofgs, everything()) %>%
    
    # Switch to long format, where vials are rows not columns
    pivot_longer(cols = 14:ncol(.), 
                 names_to = "sample_id", 
                 values_to = "sample_count") %>%
    
    # remove sample replicate numbers
    mutate(sample_id = str_replace(sample_id, "\\.[0-9]+$", ""))
  
    # One can remove all rows with no identifications using the following code, 
    # however having the rows with no IDs is useful for determining and visualizing species presence,
    # so this code was left unused:
    # filter(sample_count != 0)
  
  return(cleaned)
}
```

# Load in raw data and clean
```{r message=FALSE}
# Some original downloads were edited due to errors in copying 1s/Is and Os/0s
# Plyr package is required for ldply(), but overwrites functions that are needed later on
library(plyr)


# Metadata file should be manually updated with each batch
AQsampleIDs <- read_csv("C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_sample_ids.csv") %>% clean_names()

# Each batch folder should contain only the read tables from JonahVentures. ESV tables should not be included.
folder1470_AL <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1470_AL"
JVB1470_AL <- eDNA_clean(folder1470_AL)

folder1470_AN <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1470_AN"
JVB1470_AN <- eDNA_clean(folder1470_AN)

folder1470_AR <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1470_AR"
JVB1470_AR <- eDNA_clean(folder1470_AR)

# Unknown parsing issues below. Seven files in total are affected.
folder1554 <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1554"
JVB1554 <- eDNA_clean(folder1554)

folder1703_july <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1703_july"
JVB1703_july <- eDNA_clean(folder1703_july)

folder1703_nov <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB1703_nov"
JVB1703_nov <- eDNA_clean(folder1703_nov)

folder2116 <- "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/raw_data/eDNA_downloads/JVB2116"
JVB2116 <- eDNA_clean(folder2116)

detach("package:plyr")
```

# Bind eDNA batches
```{r}
# Combine all eDNA
all_eDNA <- bind_rows(JVB1470_AL, JVB1470_AN, JVB1470_AR, JVB1554, JVB1703_july, JVB1703_nov, JVB2116) %>%
  
  # Join sample information to IDs
  left_join(AQsampleIDs, by = c("sample_id" = "barcode")) %>%
  
  # Format dates and factors
  mutate(date = mdy(date),
         # Contiguous slough sites are ordered first in order of their distance (with barriers) from the ocean.
         # Vernal pool sites and then miscellaneous samples are at the end of the site list.
         site = factor(site, levels = c("MO1", "PIER", "CUL1", "NVBR", "NMC", "NEC", 
                                        "NPB", "NPB1", "NPB2", "NWP", "NDC",
                                        "VP1", "NVP2", "NVP4", 
                                        "bird dropping", "owl pellet", "pollen", "soil amendment"))) %>%
  
  # Remove rows without species IDs
  filter(kpcofgs != "", sample_count != 0) %>%
  
  # Rename primer column
  rename(primer = TestId)
```

# Get invertebrate eDNA
```{r}
# Subset eDNA to only include aquatic invert eDNA
aqinv_eDNA <- all_eDNA %>%
  
  # Remove irrelevant samples
  filter(!site %in% c("pollen", "soil amendment"),
         
         # Select invert taxa
         Phylum %in% c("Arthropoda", "Mollusca", "Annelida", "Nematozoa", "Platyhelminthes"))

# write clean table with aquatic invert eDNA
write_csv(aqinv_eDNA, "C:/Users/Admin/Documents/Lab Work/CCBER/Aquatic-Invertebrates/cleaned_data/eDNA_tables/aqinv_eDNA.csv")

# write clean table with aquatic invert eDNA from only vernal pools
aqinv_eDNA %>% 
  
  # select only vernal pool data
  filter(site %in% c("VP1", "NVP2", "NVP4")) %>% 
  write_csv("cleaned_data/eDNA_tables/vernal_pool_aqinv_eDNA.csv")

rm(JVB1470_AL, JVB1470_AN, JVB1470_AR, JVB1554, JVB1703_july, JVB1703_nov, JVB2116, folder1470_AL, folder1470_AN, folder1470_AR, folder1554, folder1703_july, folder1703_nov, folder2116)
```

# Create summary tables

## Get aquatic invertebrate species list
```{r}
AQspeciesList <- aqinv_eDNA %>% 
  distinct(kpcofgs, .keep_all = T) %>% 
  select(kpcofgs:percent_match)
```

## Summarize taxa by site and primer
```{r}
# get counts of taxa identified for each primer and site
site_primer_taxa_count <- aqinv_eDNA %>% distinct(site, ESVId, .keep_all = T) %>% 
  group_by(site, primer) %>%
  summarize(taxa_count = n()) %>% write_csv("cleaned_data/eDNA_tables/site_primer_taxa_count.csv")
```

## List findings by primer
```{r message=FALSE}
# This chunk takes awhile to run!

# Get count of IDs from each primer
all_eDNA %>% count(primer)

# Set database access key (needs to be updated with your personal key from NCBI)
set_entrez_key('b368c15b5cae247eccf0efb48e9cdb260f09')

# Create function to search for common names
sci2comm_concat <- function(x){
  temp <- sci2comm(x, db = "ncbi")
  if(length(temp) == 0) NA else paste0(temp[[1]], collapse = ", ")
}

# summarize MiFishU primer findings
fish_MiFishU <- all_eDNA %>% filter(primer == "MiFishU") %>%
  group_by(site, kpcofgs) %>% 
  summarize(samples_sent = n(),
            matches = sum(sample_count),
            percent_match = mean(percent_match),
            common_name = sci2comm_concat(kpcofgs)) %>%
  write_csv("cleaned_data/eDNA_tables/fish_MiFishU_eDNA.csv")

# summarize ArthCOI primer findings
inverts_ArthCOI <- all_eDNA %>% filter(primer == "ArthCOI") %>% 
  group_by(site, kpcofgs) %>% 
  summarize(samples_sent = n(),
            matches = sum(sample_count),
            percent_match = mean(percent_match),
            common_name = sci2comm_concat(kpcofgs)) %>%
  write_csv("cleaned_data/eDNA_tables/inverts_ArthCOI_eDNA.csv")

# summarize 23S, 16S, and 18S primer findings
microbes_16_18_23 <- all_eDNA %>% filter(primer %in% c("23", "18", "16")) %>% 
  group_by(site, kpcofgs) %>% 
  summarize(samples_sent = n(),
            matches = sum(sample_count),
            percent_match = mean(percent_match),
            common_name = sci2comm_concat(kpcofgs)) %>%
  write_csv("cleaned_data/eDNA_tables/microbes_16_18_23_eDNA.csv")

# summarize ITS primer findings
fungi_ITS <- all_eDNA %>% filter(primer == "ITS") %>% 
  group_by(site, kpcofgs) %>% 
  summarize(samples_sent = n(),
            matches = sum(sample_count),
            percent_match = mean(percent_match),
            common_name = sci2comm_concat(kpcofgs)) %>%
  write_csv("cleaned_data/eDNA_tables/fungi_ITS_eDNA.csv")
```

# Vizualize data
```{r automatic chart of taxa count}
ggplot(site_primer_taxa_count) + geom_bar(aes(x = site, y = taxa_count, fill = primer), stat = "identity") +
  ggtitle("eDNA Identified Taxa by Site and Primer") + ylab("taxa count") + 
  theme(axis.text.x = element_text(angle = -45, vjust = -1))
```

```{r manual chart of aquatic taxa count}
ggplot(aqinv_eDNA, aes(x = site)) + geom_bar(stat = "count") + ggtitle("Unique DNA Barcode Observations by Site") +
  theme(axis.text.x = element_text(angle = -45, vjust = -1))

ggplot(aqinv_eDNA, aes(x = ncos_or_copr)) + geom_bar(stat = "count") + ggtitle("Unique DNA Barcode Observations by Reserve")

taxa_comparison <- aqinv_eDNA %>% group_by(site) %>% 
  summarize(Phylum = n_distinct(Phylum), Class = n_distinct(Class),
                   Order = n_distinct(Order), Family = n_distinct(Family),
                   Species = n_distinct(Species)) %>%
  pivot_longer(2:6, names_to = "taxa", values_to = "count") %>%
  mutate(taxa = factor(taxa, levels = c("Phylum", "Class", "Order", "Family", "Species")))

taxaColors <- c("#A86632", "#6B4026", "#B64E38", "#1F1F1F")

taxa_comparison %>% ggplot(aes(x = taxa, y = count, fill = site)) + 
  geom_bar(stat = "identity") +
  ggtitle("Distinct Taxa Counts by Site and Source") +
  scale_fill_manual(values = c(scales::hue_pal()(14), taxaColors))
  

taxa_comparison %>% filter(!(site %in% c("bird dropping", "owl pellet", 
                                         "pollen", "soil amendment"))) %>%
  ggplot(aes(x = taxa, y = count, fill = site)) + 
  geom_bar(stat = "identity") +
  ggtitle("Distinct Taxa Counts by Aquatic Site")
```

```{r aquatic species tree diagrams, fig.height=10, fig.width=14, warning=FALSE}

heatTree <- function(input, title) {
  input %>% parse_tax_data(class_cols = 5:11) %>%
  filter_taxa(taxon_names != "") %>%
  heat_tree(node_label = taxon_names,
            node_color = n_obs,
            node_size = n_obs,
            overlap_avoidance = 20,
            node_label_size_range = c(0.021, 0.026),
            repel_force = 10,
            node_legend_title = "Taxon",
            node_color_axis_label = "# DNA matches",
            node_color_digits = 1,
            title = title,
            title_size = 0.03,
            aspect_ratio = 1.5,
            margin_size = c(0.003, 0.003, 0.003, 0.003),
            layout = "da")
}

# All samples

aqinv_eDNA %>% heatTree(" Invertebrate Species found by eDNA in the Devereux Slough, All Dates")

aqinv_eDNA %>% filter(site == "MO1") %>% 
  heatTree(" Invertebrate Species detected at Mouth of Devereux Slough, All Dates")

aqinv_eDNA %>% filter(site == "PIER") %>% 
  heatTree(" Invertebrate Species detected at COPR Pier, All Dates")

aqinv_eDNA %>% filter(site == "CUL1") %>% 
  heatTree(" Invertebrate Species detected at Culvert, All Dates")

aqinv_eDNA %>% filter(site == "NVBR") %>% 
  heatTree(" Invertebrate Species detected at Venoco Bridge, All Dates")

aqinv_eDNA %>% filter(site == "NEC") %>% 
  heatTree(" Invertebrate Species detected at Northeast Channel, All Dates")

aqinv_eDNA %>% filter(site == "NDC") %>% 
  heatTree(" Invertebrate Species detected at Devereux Creek, All Dates")

aqinv_eDNA %>% filter(site == "NWP") %>% 
  heatTree(" Invertebrate Species detected at NCOS West Pond, All Dates")

aqinv_eDNA %>% filter(site == "NPB") %>% 
  heatTree(" Invertebrate Species detected at North Phelps Bridge, All Dates")

aqinv_eDNA %>% filter(site == "NPB1") %>% 
  heatTree(" Invertebrate Species detected at North Phelps Bridge 1, All Dates")

aqinv_eDNA %>% filter(site == "NPB2") %>% 
  heatTree(" Invertebrate Species detected at North Phelps Bridge 2, All Dates")

aqinv_eDNA %>% filter(site == "VP1") %>% 
  heatTree(" Invertebrate Species detected at Vernal Pool 1, All Dates")
```

```{r nmds}
library(vegan)
library(ggrepel)

# Create presence table
presence_aqinv_eDNA <- aqinv_eDNA %>%
  filter(!site %in% c("owl pellet", "bird dropping")) %>%
  mutate(presence = 1) %>%
  pivot_wider(id_cols = site,
              names_from  = kpcofgs, 
              values_from = presence,
              values_fn = mean) %>%
  replace(is.na(.), 0)

# Create NMDS
NMDS_aqinv_eDNA <- presence_aqinv_eDNA %>%
  select(-1) %>%
  metaMDS(dist = "bray")

stressplot(NMDS_aqinv_eDNA)

# Add site type information
NCOScopr <- AQsampleIDs %>% 
  select(site, ncos_or_copr) %>%
  mutate(ncos_or_copr = case_when(site %in% c("VP1", "NVP2", "NVP4") ~ "Vernal Pool",
                                  TRUE ~ ncos_or_copr)) %>%
  unique()

NMDS_aqinv_eDNA.scores <- as.data.frame(scores(NMDS_aqinv_eDNA)$sites) %>%
  mutate(site = presence_aqinv_eDNA$site) %>%
  left_join(NCOScopr, by = "site")

ggplot(NMDS_aqinv_eDNA.scores, aes(NMDS1, NMDS2, label = site, color = ncos_or_copr)) +
  geom_mark_hull(aes(fill = ncos_or_copr), alpha = 0.2) + 
  geom_point() + geom_text_repel() + labs(color = "Site Type") 

rm(presence_aqinv_eDNA, NMDS_aqinv_eDNA, NCOScopr)
```







